src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py

```diff
@@ -0,0 +1,311 @@
+        snap_metadata = MetadataManager(self.fs, snap_metadata_path, 0o640)
+        snap_metadata.refresh()
+        snap_incarnation = snap_metadata.get_global_option(MetadataManager.GLOBAL_META_KEY_INCARNATION).encode('utf-8')
+        # subvolume snapshot data path: /volumes/_nogroup/<subvol-name>/.snap/<snap-name>/<UUID>
+        return os.path.join(snap_base_path, snap_incarnation)
+
+    def create_snapshot(self, snapname):
+        snappath = self.snapshot_path(snapname)
+        mksnap(self.fs, snappath)
+
+    def is_snapshot_protected(self, snapname):
+        try:
+            self.metadata_mgr.get_option('protected snaps', snapname)
+        except MetadataMgrException as me:
+            if me.errno == -errno.ENOENT:
+                return False
+            else:
+                log.warning("error checking protected snap {0} ({1})".format(snapname, me))
+                raise VolumeException(-errno.EINVAL, "snapshot protection check failed")
+        else:
+            return True
+
+    def has_pending_clones(self, snapname):
+        try:
+            return self.metadata_mgr.section_has_item('clone snaps', snapname)
+        except MetadataMgrException as me:
+            if me.errno == -errno.ENOENT:
+                return False
+            raise
+
+    def remove_snapshot(self, snapname):
+        if self.is_snapshot_protected(snapname):
+            raise VolumeException(-errno.EINVAL, "snapshot '{0}' is protected".format(snapname))
+        snappath = self.snapshot_path(snapname)
+        rmsnap(self.fs, snappath)
+
+    def list_snapshots(self):
+        try:
+            dirpath = os.path.join(self.base_path,
+                                   self.vol_spec.snapshot_dir_prefix.encode('utf-8'))
+            return listdir(self.fs, dirpath)
+        except VolumeException as ve:
+            if ve.errno == -errno.ENOENT:
+                return []
+            raise
+
+    def _protect_snapshot(self, snapname):
+        try:
+            self.metadata_mgr.add_section("protected snaps")
+            self.metadata_mgr.update_section("protected snaps", snapname, "1")
+            self.metadata_mgr.flush(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_INCARNATION).encode('utf-8'))
+        except MetadataMgrException as me:
+            log.warning("error updating protected snap list ({0})".format(me))
+            raise VolumeException(-errno.EINVAL, "error protecting snapshot")
+
+    def _unprotect_snapshot(self, snapname):
+        try:
+            self.metadata_mgr.remove_option("protected snaps", snapname)
+            self.metadata_mgr.flush(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_INCARNATION).encode('utf-8'))
+        except MetadataMgrException as me:
+            log.warning("error updating protected snap list ({0})".format(me))
+            raise VolumeException(-errno.EINVAL, "error unprotecting snapshot")
+
+    def protect_snapshot(self, snapname):
```

Comment:
I do not see a need to have protect/unprotect calls for v2 subvolumes.

Reply:
Removed as part of https://github.com/ceph/ceph/pull/35755